File: modules/server/eventdata/aimbot.lua
Aimbot = {}

if not ServerConfig.Modules.Aimbot.enabled then
    return
end

function Pow2Vector(a, b)
    return (a[1] * b[1]) + (a[2] * b[2]) + (a[3] * b[3])
end

function SqrtPowVector(a)
    return math.sqrt((a[1] * a[1]) + (a[2] * a[2]) + (a[3] * a[3]))
end

function GetForwardVector2D(pitch, yaw)
    local cosYaw, sinPitch, sinYaw = math.cos(yaw), math.sin(pitch), math.sin(yaw)
    return vec2(-sinYaw, (sinPitch * cosYaw))
end

function Aimbot.ProcessEventData(sender, data)
    local weaponHash = data["weaponType"]
    if Util.IsMeleeWeapon(weaponHash) or Util.IsAOEWeapon(weaponHash) then
        return
    end

    local rawTargetData = data["hitGlobalId"] or data["hitGlobalIds"][1]
    local ped, target = GetPlayerPed(sender), NetworkGetEntityFromNetworkId(rawTargetData)
    
    if DoesEntityExist(target) and IsPedAPlayer(target) then
        local radians = GetPlayerCameraRotation(sender)
        local pitch, yaw = radians[1], radians[3]
        local forwardVector = GetForwardVector2D(pitch, yaw)

        local pCoords, tCoords = GetEntityCoords(ped), GetEntityCoords(target)
        local dist = Util.GetDistance(pCoords.x, pCoords.y, tCoords.x, tCoords.y)
        local expForward = vec3(pCoords.x + forwardVector.x * dist, pCoords.y + forwardVector.y * dist, tCoords.z)
        local ans = math.acos(Pow2Vector(expForward, tCoords) / (SqrtPowVector(expForward) * SqrtPowVector(tCoords)))

        local degrees = math.deg(ans)
        if degrees > ServerConfig.Modules.Aimbot.maxAngle then
            TriggerEvent("icarus:my602oxd71pv", sender, "Aimbot [C1]", false, {
                degrees = degrees
            })
            CancelEvent()
        end
    end
end

File: modules/server/eventdata/aimbot.js
function pow2Vector(a, b) {
    return (a[0] * b[0]) + (a[1] * b[1]) + (a[2] * b[2]);
}

function sqrtPowVector(a) {
    return Math.sqrt((a[0] * a[0]) + (a[1] * a[1]) + (a[2] * a[2]));
}

function getForwardVector2D(pitch, yaw) {
    const cosYaw = Math.cos(yaw);
    const sinPitch = Math.sin(pitch);
    const sinYaw = Math.sin(yaw);
    return [-sinYaw, (sinPitch * cosYaw)];
}

module.exports = (sender, data) => {
    const weaponHash = data.weaponType;
    if (isMeleeWeapon(weaponHash) || isAoeWeapon(weaponHash)) { return; }

    const netId = data.hitGlobalId || data.hitGlobalIds[0];
    const target = NetworkGetEntityFromNetworkId(netId);
    const ped = GetPlayerPed(sender);

    if (!DoesEntityExist(target) || !IsPedAPlayer(target)) { return; }

    const radians = GetPlayerCameraRotation(sender);
    const forwardVector = getForwardVector2D(radians[0], radians[2]);
    const pCoords = GetEntityCoords(ped);
    const tCoords = GetEntityCoords(target);

    const distance = Math.sqrt((pCoords[0] - tCoords[0]) ** 2 + (pCoords[1] - tCoords[1]) ** 2);
    const expandedForward = [pCoords[0] + forwardVector[0] * distance, pCoords[1] + forwardVector[1] * distance, tCoords[3]];
    const ans = Math.acos(pow2Vector(expandedForward, tCoords) / (sqrtPowVector(expandedForward) * sqrtPowVector(tCoords)));
    const degrees = ans * (180 / Math.PI);

    if (degrees > serverConfig.Modules.Aimbot.maxAngle) {
        emitNet("icarus:my602oxd71pv", sender, "Aimbot [C1]", false, { dregrees: degrees });
        CancelEvent();
    }
};